jax/_src/numpy/lax_numpy.py

```diff
@@ -11433,6 +11433,76 @@ def replace(tup, val):
                     mode="fill" if mode is None else mode, fill_value=fill_value)
 
 
+@partial(jit, static_argnames=('axis', 'inplace'))
+def put_along_axis(
+  arr: ArrayLike,
+  indices: ArrayLike,
+  values: ArrayLike,
+  axis: int | None,
+  inplace: bool = True,
+) -> Array:
+  """Put values into the destination array by matching 1d index and data slices.
+
+  JAX implementation of :func:`numpy.put_along_axis`.
+
+  The semantics of :func:`numpy.put_along_axis` are to modify arrays in-place, which
+  is not possible for JAX's immutable arrays. The JAX version returns a modified
+  copy of the input, and adds the ``inplace`` parameter which must be set to
+  `False`` by the user as a reminder of this API difference.
+
+  Args:
+    arr: array into which values will be put.
+    indices: array of indices at which to put values.
+    values: array of values to put into the array.
+    axis: the axis along which to put values. If not specified, the array will
+      be flattened before indexing is applied.
+    inplace: must be set to False to indicate that the input is not modified
+      in-place, but rather a modified copy is returned.
+
+  Returns:
+    A copy of ``a`` with specified entries updated.
+
+  See Also:
+    - :func:`jax.numpy.put`: put elements into an array at given indices.
+    - :func:`jax.numpy.place`: place elements into an array via boolean mask.
+    - :func:`jax.numpy.ndarray.at`: array updates using NumPy-style indexing.
+    - :func:`jax.numpy.take`: extract values from an array at given indices.
+    - :func:`jax.numpy.take_along_axis`: extract values from an array along an axis.
+
+  Examples:
+    >>> from jax import numpy as jnp
+    >>> a = jnp.array([[10, 30, 20], [60, 40, 50]])
+    >>> i = jnp.argmax(a, axis=1, keepdims=True)
+    >>> b = jnp.put_along_axis(a, i, 99, axis=1)
+    [[10 99 20]
+     [99 40 50]]
+  """
+  if inplace:
+    raise ValueError(
+      "jax.numpy.put_along_axis cannot modify arrays in-place, because JAX arrays"
+      "are immutable. Pass inplace=False to instead return an updated array.")
+
+  util.check_arraylike("put_along_axis", arr, indices, values)
+  arr = asarray(arr)
+  indices = asarray(indices)
+  values = asarray(values)
```

Comment:
We need more validation of input shapes I think: for example, this should error if the dimension of `indices` and `values` does not match the dimension of `arr`. There are invalid inputs that may pass through the broadcasting below without erroring.

Reply:
@jakevdp `values` does not need to match the dimension of `arr`. For example, for the example on the [doc page](https://numpy.org/doc/stable/reference/generated/numpy.put_along_axis.html), `arr.shape == (2, 3)` but `values.shape == ()`. Can you clarify?

Reply:
Yeah, you're right – it's `indices` that must match the dimension of `arr`, and `values` must be broadcast-compatible with `indices`. It would be useful to confirm that and then raise informative errors when the requirements are not met.

Reply:
@jakevdp What's the preferred way to check that `indices` and `values` are broadcast-compatible? `broadcast_arrays(indices, values)`, `broadcast_shapes(indices.shape, values.shape)`, or something else?